crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,618 @@
+///
+///     qubits = [Qubit()]
+///     data = CircuitData(qubits)
+///     data.append(CircuitInstruction(XGate(), (qubits[0],), ()))
+///     assert(data[0] == data[0]) # => Ok.
+///     assert(data[0] is data[0]) # => PANICS!
+///
+/// .. warning::
+///
+///     This is an internal interface and no part of it should be relied upon
+///     outside of Qiskit.
+///
+/// Args:
+///     qubits (Iterable[:class:`.Qubit`] | None): The initial sequence of
+///         qubits, used to map :class:`.Qubit` instances to and from its
+///         indices.
+///     clbits (Iterable[:class:`.Clbit`] | None): The initial sequence of
+///         clbits, used to map :class:`.Clbit` instances to and from its
+///         indices.
+///     data (Iterable[:class:`.CircuitInstruction`]): An initial instruction
+///         listing to add to this container. All bits appearing in the
+///         instructions in this list must also exist in ``qubits`` and
+///         ``clbits``.
+///     reserve (int): An additional capacity to reserve beyond the container's
+///         initial capacity.
+///
+/// Raises:
+///     KeyError: if ``data`` contains a reference to a bit that is not present
+///         in ``qubits`` or ``clbits``.
+#[pyclass(sequence, module = "qiskit._accelerate.quantum_circuit")]
+#[derive(Clone, Debug)]
+pub struct CircuitData {
+    /// The interned instruction listing.
+    data: Vec<InternedInstruction>,
+    /// The intern context used to intern the instruction listing.
+    intern_context: InternContext,
+    /// The qubits registered (e.g. through :meth:`~.CircuitData.add_qubit`).
+    qubits_native: Vec<PyObject>,
+    /// The clbits registered (e.g. through :meth:`~.CircuitData.add_clbit`).
+    clbits_native: Vec<PyObject>,
+    /// Map of :class:`.Qubit` instances to their index in
+    /// :attr:`.CircuitData.qubits`.
+    qubit_indices_native: HashMap<BitAsKey, BitType>,
+    /// Map of :class:`.Clbit` instances to their index in
+    /// :attr:`.CircuitData.clbits`.
+    clbit_indices_native: HashMap<BitAsKey, BitType>,
+    /// The qubits registered, as a ``list[Qubit]``.
+    #[pyo3(get)]
+    qubits: Py<PyList>,
+    /// The clbits registered, as a ``list[Clbit]``.
+    #[pyo3(get)]
+    clbits: Py<PyList>,
+}
+
+/// A private enumeration type used to extract arguments to pymethods
+/// that may be either an index or a slice.
+#[derive(FromPyObject)]
+pub enum SliceOrInt<'a> {
+    Slice(&'a PySlice),
+    Int(isize),
+}
+
+#[pymethods]
+impl CircuitData {
+    #[new]
+    #[pyo3(signature = (qubits=None, clbits=None, data=None, reserve=0))]
+    pub fn new(
+        py: Python<'_>,
+        qubits: Option<&PyAny>,
+        clbits: Option<&PyAny>,
+        data: Option<&PyAny>,
+        reserve: usize,
+    ) -> PyResult<Self> {
+        let mut self_ = CircuitData {
+            data: if reserve > 0 {
+                Vec::with_capacity(reserve)
+            } else {
+                Vec::new()
+            },
```

Comment:
Yeah, you're right! The [Guarantees](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) section makes this clear.

I'll change this. When I read the docs for both of these originally, I'd assumed that `Vec::new` would allocate some initial default capacity, since the docs for `with_capacity` mention explicitly that a value of `0` avoids allocation.

Reply:
Fixed in 65aa628.